#' Draw a blank plot
#'
#' @noRd
blank_plot <- function(){
  graphics::plot(0,
       type = 'n',
       axes = FALSE,
       ann = FALSE)
}

#' Plot the continuous phenotype as a phylogenetic tree filled in with a color
#' gradient
#'
#' @description Plot the continuous phenotype on tree with the phenotype
#'   coloring the branches.
#'
#' @param tr Phylo
#' @param pheno_vector Vector. Length = Ntip(tr). Names(pheno_vector) ==
#'   tr$tip.label.
#' @param pheno_anc_rec Vector. Length = Nnode(tr).
#' @param tr_type Characer. Default = "phylogram". User can supply either:
#'   "phylogram" or "fan." Determines how the trees are plotted in the output.
#' @param strain_key Matrix. Default = NULL.
#'
#' @return A tree plot where the tree edges are filled in with colors
#'   corresponding to the phenotypic trait values.
#'
#' @noRd
plot_continuous_phenotype <- function(tr,
                                      pheno_vector,
                                      pheno_anc_rec,
                                      tr_type,
                                      strain_key = NULL){
  # Check inputs ---------------------------------------------------------------
  check_tree_is_valid(tr)
  check_equal(length(pheno_vector), ape::Ntip(tr))
  check_equal(length(pheno_anc_rec), ape::Nnode(tr))
  if (!identical(names(pheno_vector), tr$tip.label)) {
    stop("Phenotype vector must have same names as tree tips.")
  }
  check_is_string(tr_type)

  # Function -------------------------------------------------------------------
  plot_p_recon <- phytools::contMap(tr,
                                    pheno_vector,
                                    method = "user",
                                    anc.states = pheno_anc_rec,
                                    plot = FALSE)

  line_width <- 3
  if (ape::Ntip(tr) >= 200) {
    line_width <- 1
  }

  if (tr_type == "phylogram") {
    # phylogram version
    graphics::plot(plot_p_recon,
         add = TRUE,
         ylim = c(-1 / 25 * ape::Ntip(tr), ape::Ntip(tr)),
         colors = plot_p_recon$cols,
         lwd = line_width,
         ftype = "off",
         type = "phylogram",
         offset = 1.7,
         outline = FALSE)
  } else {
    # fan version
    graphics::plot(plot_p_recon,
                   add = TRUE,
                   colors = plot_p_recon$cols,
                   lwd = 1,
                   type = "fan",
                   ftype = "off",
                   outline = FALSE)
  }

  if (!is.null(strain_key)) {
    strain_color_key <- assign_colors_to_strain_types(tr, strain_key)
    ape::tiplabels(tip = c(1:ape::Ntip(tr)),
                   pch =  19, # circle
                   cex = .55, # dot size
                   col = unlist(strain_color_key[, 2]))

    graphics::legend("left",
                     bty = "o",
                     legend = unique(strain_color_key[, 1]),
                     col = unique(strain_color_key[, 2]),
                     lty = 1,
                     ncol = 1,
                     lwd = 1)
  }

}


#' Draw histogram of the absolute value of phenotype change on each high
#' confidence tree edge
#'
#' @description Plot a histogram to show the change in phenotype on each tree
#'   edge. Plot phenotype change as absolute value change. Exclude low
#'   confidence tree edges.
#'
#' @noRd
#' @param all_trans List of 2 lists. $observed_pheno_non_trans_delta and
#'   $observed_pheno_trans_delta. Position in each corresponds to one genotype.
#'   Contains informatino about amount of phenotype change.
#' @param tr Phylo.
#' @param index Integer. Indexes which genotype is being plotted.
#' @param non_trans_color Character. Color of non-transition histogram bars.
#' @param trans_color Character. Color of transition histogram bars.
#'
#' @return Histogram showing the change in phenotype on each high confidence
#'   tree edge. Transition edges are different color than non-transition edges.
#'   Low confidence edges are excluded. Change in phenotype is absolute value.
hist_abs_hi_conf_delta_pheno <- function(all_trans,
                                         tr,
                                         index,
                                         non_trans_color,
                                         trans_color){
  # Check inputs ---------------------------------------------------------------
  check_is_number(index)
  if (index > all_trans$num_genotypes) {
    stop("Index must correspond to one of the genotypes")
  }
  check_for_root_and_bootstrap(tr)
  check_is_string(non_trans_color)
  check_is_string(trans_color)
  if (non_trans_color == trans_color) {
    stop("These tree edges need to be different colors")
  }

  # Function -------------------------------------------------------------------
  hist_cex_size <- 1
  graphics::par(mar = c(4, 4, 4, 4))
  graphics::hist(all_trans$observed_pheno_non_trans_delta[[index]],
                 font = 1,
       col = non_trans_color,
       ylab = "Count",
       xlab = expression(paste("|", Delta, "| Phenotype")),
       main = expression(paste("|", Delta, "| Phenotype Per Edge")),
       xlim = c(0,
                1.2 * max(all_trans$observed_pheno_trans_delta[[index]],
                          all_trans$observed_pheno_non_trans_delta[[index]])),
       ylim = c(0,
                0.75 * sum(length(all_trans$observed_pheno_trans_delta[[index]]),
                           length(all_trans$observed_pheno_non_trans_delta[[index]]))),
       cex.main = hist_cex_size,
       cex.lab = hist_cex_size,
       cex.axis = hist_cex_size)

  graphics::hist(all_trans$observed_pheno_trans_delta[[index]],
       # breaks = ape::Nedge(tr) / 4,
       col = trans_color,
       add = TRUE)

  graphics::legend("topright",
         title = "Hi Conf. Geno. Edge Type",
         legend = c("Transition", "Non-transition"),
         col = c(trans_color, non_trans_color),
         pch = 15,
         bty = "o",
         cex = hist_cex_size)
}

#' Draw Manhattan Plot
#'
#' @description Create a Manhattan plot of GWAS hit p-values.
#'
#' @noRd
#' @param geno_pheno_name Character. Will be printed in title of manhattan plot.
#' @param pval_hits Vector. P-values for all loci.
#' @param fdr Number. False discorvery rate.
#' @param test_name Character. Either "continuous", "synchronous", or "phyc".
#'   Will be printed in title of Manhattan plot.
#'
#' @return A Manhattan plot of all of the hit values.
make_manhattan_plot <- function(geno_pheno_name,
                                pval_hits,
                                fdr,
                                test_name){
  # Check inputs ---------------------------------------------------------------
  check_is_string(geno_pheno_name)
  check_is_number(fdr)
  check_str_is_test_name(test_name)

  # Function -------------------------------------------------------------------
  transparent_grey <- grDevices::rgb(0, 0, 0, 0.25)

  # Make test names prettier
  if (test_name == "continuous") {
    test_name <- "Continuous Test:"
  } else if (test_name == "synchronous") {
    test_name <- "Synchronous Test:"
  }  else {
    test_name <- "PhyC Test:"
  }

  manhattan_cex <- 1
  neg_log_p_value <- data.frame(pval_hits)
  check_is_number(neg_log_p_value[1, 1])
  neg_log_p_with_num <- cbind(1:nrow(neg_log_p_value), neg_log_p_value)
  colnames(neg_log_p_with_num)[1] <- "Locus Significance"
  ymax <- max(-log(0.01), neg_log_p_with_num[, 2, drop = TRUE])
  with(neg_log_p_with_num,
       graphics::plot(x = neg_log_p_with_num[, 1],
            y = jitter(neg_log_p_with_num[, 2, drop = TRUE]),
            type = "p",
            cex.main = manhattan_cex,
            cex.lab = manhattan_cex,
            cex.axis = manhattan_cex,
            main = paste(test_name, geno_pheno_name, sep = " "),
            col = transparent_grey,
            pch = 19,
            xaxt = 'n',
            xlab = "Genetic loci",
            ylim = c(0, ymax),
            ylab = "-ln(FDR Corrected P-value)"))

  graphics::abline(h = fdr,
                   col = "red")
  graphics::legend("topright",
         bty = "n",
         legend = "Significance Threshold",
         col = "red",
         pch = "-",
         cex = manhattan_cex)
}

#' Plot tree with specific edges colored
#'
#' @noRd
#' @param tr Phylo.
#' @param edges_to_highlight List of vectors.
#' @param geno_confidence List of vectors.
#' @param edge_color_na. Character. Color.
#' @param edge_color_bright Character. Color.
#' @param title Character. Plot title.
#' @param trans_or_recon Character. Either "recon" or "trans."
#' @param index Number.
#' @param legend_baseline Character. Legend name for black lines.
#' @param legend_highlighted Character. Legend name for highlighted lines.
#' @param tr_type Characer. Default = "phylogram". User can supply either:
#'   "phylogram" or "fan." Determines how the trees are plotted in the output.
#' @param strain_type Matrix or NULL
#' @param tip_name_log Logical. TRUE == plot tree with tip names.
#' @return Plot of a tree with certain edges colored.
plot_tr_w_color_edges <- function(tr,
                                  edges_to_highlight,
                                  geno_confidence,
                                  edge_color_na,
                                  edge_color_bright,
                                  title,
                                  trans_or_recon,
                                  index,
                                  legend_baseline,
                                  legend_highlighted,
                                  tr_type,
                                  strain_key,
                                  tip_name_log){
  # Check input ----------------------------------------------------------------
  check_for_root_and_bootstrap(tr)
  check_is_string(edge_color_na)
  check_is_string(edge_color_bright)
  if (edge_color_na == edge_color_bright) {
    stop("These tree edges need to be different colors")
  }
  check_is_string(legend_baseline)
  check_is_string(legend_highlighted)
  check_is_number(index)
  if (index > length(edges_to_highlight)) {
    stop("Index must select an item from edges_to_highlight")
  }
  check_equal(length(geno_confidence), length(edges_to_highlight))
  check_equal(length(geno_confidence[[index]]), ape::Nedge(tr))
  check_is_string(trans_or_recon)
  if (!trans_or_recon %in% c("recon", "trans")) {
    stop("String must be trans or recon")
  }
  check_is_string(title)
  check_is_string(tr_type)
  if (!is.null(strain_key)) {
    check_dimensions(strain_key,
                     min_rows = 1,
                     exact_cols = 1,
                     min_cols = 1)
  }
  check_class(tip_name_log, "logical")

  # Function -------------------------------------------------------------------
  tree_legend_cex <- 0.5
  edge_color_baseline <- "black"
  edge_color <- rep(edge_color_baseline, ape::Nedge(tr))
  if (trans_or_recon == "recon") {
    check_equal(length(edges_to_highlight[[index]]), ape::Nedge(tr))
    edge_color[edges_to_highlight[[index]] == 1] <- edge_color_bright
  } else if (trans_or_recon == "trans") {
    check_equal(length(edges_to_highlight[[index]]$transition), ape::Nedge(tr))
    edge_color[edges_to_highlight[[index]]$transition == 1] <- edge_color_bright
  }
  edge_color[geno_confidence[[index]] == 0] <- edge_color_na # grey out long
  # edges and low ML bootstrap support

  edge_width <- 1
  if (ape::Ntip(tr) >= 200) {
    edge_width <- 0.25
  }

  edge_length_log <- TRUE
  if (tr_type == "phylogram") {
    edge_length_log <- FALSE
  }

  graphics::par(mar = c(4, 4, 4, 4))
  graphics::plot(tr,
                 type = tr_type,
                 font = 1,
                 show.tip.label = tip_name_log,
                 edge.color = edge_color,
                 main = title,
                 use.edge.length = edge_length_log,
                 label.offset = 0.25,
                 adj = 0,
                 cex = tree_legend_cex,
                 edge.width = edge_width)

  if (!is.null(strain_key)) {
    strain_color_key <- assign_colors_to_strain_types(tr, strain_key)
    ape::tiplabels(tip = c(1:ape::Ntip(tr)),
                   pch =  19, # circle
                   cex = .55, # dot size
                   col = unlist(strain_color_key[, 2]))

    graphics::legend("left",
                     bty = "o",
                     legend = unique(strain_color_key[, 1]),
                     col = unique(strain_color_key[, 2]),
                     lty = 1,
                     ncol = 1,
                     lwd = 1,
                     cex = tree_legend_cex)
  }

  graphics::legend("topleft",
                   bty = "o",
                   legend = c(legend_baseline,
                              legend_highlighted,
                              "Low confidence"),
                   col = c(edge_color_baseline,
                           edge_color_bright,
                           edge_color_na),
                   lty = 1,
                   ncol = 1,
                   lwd = 1,
                   cex = tree_legend_cex)
}

#' Plot all PhyC results
#'
#' @noRd
#' @param tr Phylo.
#' @param dir Directory where to save plots.
#' @param name Prefix in plot file name.
#' @param fdr Numeric. False discovery rate. Between 0 and 1.
#' @param num_perm Numeric. Number of permutations.
#' @param recon_hit_vals Dataframe. Nrows = number of genotypes. Ncol = 1.
#'   Corrected p-values for each genotype tested.
#' @param p_recon_edges Vector. Length = Nedge(tree). Reconstruction of
#'   phenotype.
#' @param recon_perm_obs_results List of many results. $hit_pvals. Character.
#'   P-val for each genotype. Length = number of tested genotypes.
#'   $permuted_count. List of vectors. 1 vector for each tested genotype. Length
#'   of each vector = number of permuations. $observed_overlap. Integer vector.
#'   Length = number of tested genotypes.
#' @param tr_and_pheno_hi_conf Vector of logicals. TRUE = high confidence. FALSE
#'   = low confidence. Length = Nedge(tree).
#' @param geno_confidence List of vectors. Length of individual vector =
#'   Nedge(tree). Genotype high confidence edges. Either 1 (high confidence) or
#'   0 (low confidence).
#' @param g_trans_edges List of vectors. Length of individual vector =
#'   Nedge(tree). Genotype transition edges. Either 1 (transition) or 0 (no
#'   transition).
#' @param p_trans_edges Vector. Length = Nedge(tree). Transitions marked as 1,
#'   not transition marked as 0.
#' @param snp_in_gene Either NULL or table of integers where each entry
#'   corresponds to one genotype.
#' @param tr_type Characer. Default = "phylogram". User can supply either:
#'   "phylogram" or "fan." Determines how the trees are plotted in the output.
#' @param strain_key Matrix
#'
#' @return  Plots printed into one pdf.
plot_phyc_results <- function(tr,
                              dir,
                              name,
                              fdr,
                              num_perm,
                              recon_hit_vals,
                              p_recon_edges,
                              recon_perm_obs_results,
                              tr_and_pheno_hi_conf,
                              geno_confidence,
                              g_trans_edges,
                              p_trans_edges,
                              snp_in_gene,
                              prefix,
                              grouped_logical,
                              tr_type,
                              strain_key = NULL){
  # Check input ----------------------------------------------------------------
  check_for_root_and_bootstrap(tr)
  check_if_dir_exists(dir)
  check_is_string(name)
  check_is_number(fdr)
  check_if_permutation_num_valid(num_perm)
  if (ncol(recon_hit_vals) != 1 |
      nrow(recon_hit_vals) != length(geno_confidence)) {
    stop("Dimensions of hit p-values dataframe are incorrect.")
  } # Don't change to check_dimensions because input is dataframe, not matrix.
  check_equal(length(p_recon_edges), ape::Nedge(tr))
  if (!is.null(snp_in_gene)) {
    if (!is_this_class(snp_in_gene, "table") | typeof(snp_in_gene) != "integer") {
      stop("snp_in_gene should be a table of integers")
    }
  }
  check_equal(length(p_trans_edges), ape::Nedge(tr))
  check_equal(length(geno_confidence[[1]]), ape::Nedge(tr))
  check_equal(length(g_trans_edges[[1]]), ape::Nedge(tr))
  check_if_binary_vector(geno_confidence[[1]])
  check_if_binary_vector(p_trans_edges)
  check_if_binary_vector(g_trans_edges[[1]])
  check_equal(length(tr_and_pheno_hi_conf), ape::Nedge(tr))
  check_equal(length(recon_perm_obs_results$permuted_count[[1]]), num_perm)
  check_class(recon_perm_obs_results$hit_pvals, "character")
  check_class(recon_perm_obs_results$observed_overlap, "integer")
  check_class(grouped_logical, "logical")
  check_str_is_test_name(prefix)
  check_is_string(tr_type)

  # Function -------------------------------------------------------------------

  # Set default values
  image_width <- 250
  hist_cex_size <- 1
  # File name
  if (grouped_logical) {
    fname <- paste0(dir, "/hogwash_", prefix, "_grouped_", name, ".pdf")
  } else {
    fname <- paste0(dir, "/hogwash_", prefix, "_", name, ".pdf")
  }

  transparent_red <- grDevices::rgb(1, 0, 0, 0.25)
  transparent_teal <- grDevices::rgb(0.4, 0.6, 1, alpha = 0.5, )

  # Manhattan Plot
  grDevices::pdf(fname, width = 8.5, height = 11)
  graphics::par(mfrow = c(1, 1), mar = c(5, 5, 5, 5))
  make_manhattan_plot(name, recon_hit_vals, fdr, prefix)

  # Prep heatmap
  g_trans_mat <- matrix(0, nrow = ape::Nedge(tr), ncol = length(g_trans_edges))

  for (i in 1:length(g_trans_edges)) {
    g_trans_mat[, i] <- g_trans_edges[[i]]
    g_trans_mat[geno_confidence[[i]] == 0, i] <- NA
  }

  p_recon_edges[tr_and_pheno_hi_conf == 0] <- -1
  p_mat <- matrix(p_recon_edges, nrow = length(p_recon_edges), ncol = 1)
  colnames(p_mat) <- "Edge Type"
  phenotype_annotation <- as.data.frame(p_mat)
  row.names(phenotype_annotation) <- 1:nrow(phenotype_annotation)

  temp_g_trans_mat <- cbind(phenotype_annotation, g_trans_mat)
  g_trans_mat <- temp_g_trans_mat[order(temp_g_trans_mat[, 1],
                                        na.last = FALSE,
                                        decreasing = FALSE ),
                                  2:ncol(temp_g_trans_mat),
                                  drop = FALSE]
  colnames(g_trans_mat) <- row.names(recon_hit_vals)

  cell_width_value <- image_width / ncol(g_trans_mat)

  significant_loci <-
    data.frame("Locus Significance" = rep("Not Significant", ncol(g_trans_mat)),
                                 stringsAsFactors = FALSE)
  row.names(significant_loci) <- row.names(recon_hit_vals)
  log_p_value <- data.frame(recon_hit_vals)
  significant_loci[log_p_value > fdr] <- "Significant"
  if (!is.null(snp_in_gene)) {
    snp_in_gene <- as.data.frame(snp_in_gene, row.names = 1)
    colnames(snp_in_gene) <- "Variants in Group"
    snp_in_gene <-
      snp_in_gene[row.names(snp_in_gene) %in% row.names(log_p_value), ,
                  drop = FALSE]
    column_annot <- cbind(significant_loci, log_p_value, snp_in_gene)
  } else {
    column_annot <- cbind(significant_loci, log_p_value)
  }

  ordered_by_p_val <-
    g_trans_mat[,
                 match(row.names(log_p_value)[order(log_p_value[, 1])],
                       colnames(g_trans_mat)),
                 drop = FALSE]
  column_annot_ordered_by_p_val <-
    column_annot[match(row.names(log_p_value)[order(log_p_value[, 1])],
                        row.names(column_annot)), , drop = FALSE]

  if (ncol(column_annot_ordered_by_p_val) == 2) {
    colnames(column_annot_ordered_by_p_val) <- c("Locus Significance",
                                                 "-ln(FDR Corrected P-value)")
  } else {
    colnames(column_annot_ordered_by_p_val) <- c("Locus Significance",
                                                 "-ln(FDR Corrected P-value)",
                                                 "Variants in Group")
  }

  ann_colors <-
    list(`Locus Significance` = c(`Not Significant` = "white",
                                  Significant = "blue"),
         `Edge Type` = c(`Low Confidence` = "grey",
                                          `Phenotype Absence or Genotype Non-Transition` = "white",
                                          `Phenotype Presence` = "red",
                                          `Genotype Transition` = "black"))

  phenotype_annotation[phenotype_annotation == -1] <- "Low Confidence"
  phenotype_annotation[phenotype_annotation == 0] <- "Phenotype Absence or Genotype Non-Transition"
  phenotype_annotation[phenotype_annotation == 1] <- "Phenotype Presence"

  plotting_logical <- check_if_g_mat_can_be_plotted(ordered_by_p_val)
  # End heatmap prep

  # Plot tree edges vs genotype non/transitions
  if (plotting_logical) {
    colnames(ordered_by_p_val) <- substr(colnames(ordered_by_p_val), 1, 20)

    pheatmap::pheatmap(
      ordered_by_p_val,
      main = "PhyC Test: Rows=Tree Edges, Columns=Genotypes",
      cluster_cols  = TRUE,
      legend = FALSE,
      na_col = "grey",
      cluster_rows  = FALSE,
      show_rownames = FALSE,
      show_colnames = FALSE,
      color = c("white", "black"),
      angle_col = 45,
      annotation_col = column_annot_ordered_by_p_val,
      annotation_row = phenotype_annotation,
      annotation_colors = ann_colors,
      fontsize = 8,
      cellwidth = cell_width_value)
  }
  # Plot Phenotype Presence/Absence on Tree
  pheno_as_list <- list(p_recon_edges)
  pheno_conf_as_list <- list(tr_and_pheno_hi_conf)
  graphics::par(mfrow = c(1, 1),
                mgp = c(3, 1, 0),
                oma = c(0, 0, 4, 0),
                mar = c(4, 4, 4, 4))
  plot_tr_w_color_edges(tr,
                        pheno_as_list,
                        pheno_conf_as_list,
                        "grey",
                        "red",
                        paste0("\n Phenotype Reconstruction"),
                        "recon",
                        1,
                        "Wild type",
                        "Variant",
                        tr_type,
                        strain_key = strain_key,
                        tip_name_log = TRUE)

  # Prep data to report p-value rank
  rank_mat <- recon_hit_vals
  rank_mat[rank_mat == 0] <- NA
  genotype_rank_order <- rank(1 / rank_mat[, 1], na.last = TRUE, ties.method = "first")

  # Loop through significant hits:
  for (j in genotype_rank_order) {
    if (recon_hit_vals[j, 1] > fdr) {
      graphics::par(mfrow = c(1, 1),
                    mgp = c(3, 1, 0),
                    oma = c(0, 0, 4, 0),
                    mar = c(4, 4, 4, 4))
      # Plot Genotype Transitions on Tree
      plot_tr_w_color_edges(tr,
                            g_trans_edges,
                            geno_confidence,
                            "grey",
                            "red",
                            paste0(row.names(recon_hit_vals)[j],
                                   "\nGenotype Transitions"),
                            "recon",
                            j,
                            "No transition",
                            "Transition",
                            tr_type,
                            strain_key = NULL,
                            tip_name_log = TRUE)

      # Plot Null Distribution Histogram
      max_x <- 1.2 * max(recon_perm_obs_results$permuted_count[[j]],
                   recon_perm_obs_results$observed_overlap[j])

      p_val_formated <- formatC(recon_hit_vals[j, 1], format = "e", digits = 1)
      p_val_rank_formated <- rank(1 / rank_mat,
                                  na.last = TRUE,
                                  ties.method = "average")[j]
      title_line_one <- row.names(recon_hit_vals)[j]
      title_line_two <- bquote(paste("-ln(FDR Corrected P-value) = ",
                                    .(p_val_formated),
                                    " P-value rank = ",
                                    .(p_val_rank_formated)))
      graphics::hist(recon_perm_obs_results$permuted_count[[j]],
                     xlim = c(0, max_x),
                     col = transparent_teal,
                     ylab = "Count",
                     xlab = expression("N"["PhyC"]),
                     main = title_line_one)
      graphics::abline(v = recon_perm_obs_results$observed_overlap[j],
                       col = transparent_red,
                       lwd = 4)
      graphics::mtext(title_line_two, side = 3, cex = .6 * hist_cex_size)
      graphics::legend("topleft",
                       legend = c("Null", "Observed"),
                       col = c(transparent_teal, transparent_red),
                       pch = 15,
                       bty = "n",
                       cex = hist_cex_size)
    }
  }

  grDevices::dev.off()
}

#' Plot Synchronous Test results
#'
#' @noRd
#' @param tr Phylo.
#' @param dir Directory where to save plots.
#' @param name Prefix in plot file name.
#' @param fdr Numeric. False discovery rate. Between 0 and 1.
#' @param num_perm Numeric. Number of permutations.
#' @param trans_hit_vals Dataframe. Nrows = number of genotypes. Ncol = 1.
#'   Corrected p-values for each genotype tested.
#' @param trans_perm_obs_results List of many results.  $hit_pvals. Character.
#'   P-val for each genotype. Length = number of tested genotypes.
#'   $permuted_count. List of vectors. 1 vector for each tested genotype. Length
#'   of each vector = number of permuations. $observed_overlap. Integer vector.
#'   Length = number of tested genotypes.
#' @param tr_and_pheno_hi_conf Vector of logicals. TRUE = high confidence. FALSE
#'   = low confidence. Length = Nedge(tree).
#' @param geno_confidence List of vectors. Length of individual vector =
#'   Nedge(tree). Genotype high confidence edges. Either 1 (high confidence) or
#'   0 (low confidence).
#' @param g_trans_edges List of vectors. Length of individual vector =
#'   Nedge(tree). Genotype transition edges. Either 1 (transition) or 0 (no
#'   transition).
#' @param p_trans_edges Vector. Length = Nedge(tree). Transitions marked as 1,
#'   not transition marked as 0.
#' @param snp_in_gene Either NULL or Table of integers where each entry
#'   corresponds to one genotype.
#' @param tr_type Characer. Default = "phylogram". User can supply either:
#'   "phylogram" or "fan." Determines how the trees are plotted in the output.
#' @param strain_key Matrix.
#'
#'  @return  Plots printed into one pdf.
plot_synchronous_results  <- function(tr,
                                 dir,
                                 name,
                                 fdr,
                                 num_perm,
                                 trans_hit_vals,
                                 trans_perm_obs_results,
                                 tr_and_pheno_hi_conf,
                                 geno_confidence,
                                 g_trans_edges,
                                 p_trans_edges,
                                 snp_in_gene,
                                 prefix,
                                 grouped_logical,
                                 tr_type,
                                 strain_key = NULL){
  # Check input ----------------------------------------------------------------
  check_for_root_and_bootstrap(tr)
  check_if_dir_exists(dir)
  check_is_string(name)
  check_is_number(fdr)
  check_if_permutation_num_valid(num_perm)
  if (ncol(trans_hit_vals) != 1 |
      nrow(trans_hit_vals) != length(geno_confidence)) {
    stop("Dimensions of hit p-values dataframe are incorrect.")
  }
  if (!is.null(snp_in_gene)) {
    if (!is_this_class(snp_in_gene, "table")  | typeof(snp_in_gene) != "integer") {
      stop("snp_in_gene should be a table of integers")
    }
  }
  check_equal(length(p_trans_edges), ape::Nedge(tr))
  check_equal(length(geno_confidence[[1]]), ape::Nedge(tr))
  check_equal(length(g_trans_edges[[1]]), ape::Nedge(tr))
  check_if_binary_vector(geno_confidence[[1]])
  check_if_binary_vector(p_trans_edges)
  check_if_binary_vector(g_trans_edges[[1]])
  check_equal(length(tr_and_pheno_hi_conf), ape::Nedge(tr))
  check_equal(length(trans_perm_obs_results$permuted_count[[1]]), num_perm)
  check_class(trans_perm_obs_results$hit_pvals, "character")
  check_class(trans_perm_obs_results$observed_overlap, "integer")
  check_str_is_test_name(prefix)
  check_class(grouped_logical, "logical")
  check_is_string(tr_type)

  # Function -------------------------------------------------------------------
  # Set default values
  image_width <- 250
  hist_cex_size <- 1
  transparent_red <- grDevices::rgb(1, 0, 0, 0.25)
  transparent_teal <- grDevices::rgb(0.4, 0.6, 1, alpha = 0.5, )

  # File name
  if (grouped_logical) {
    fname <- paste0(dir, "/hogwash_", prefix, "_grouped_", name, ".pdf")
  } else {
    fname <- paste0(dir, "/hogwash_", prefix, "_", name, ".pdf")
  }

  # Start Plotting
  grDevices::pdf(fname, width = 8.5, height = 11)

  # Manhattan Plot
  graphics::par(mfrow = c(1, 1), mar = c(5, 5, 5, 5))
  make_manhattan_plot(name, trans_hit_vals, fdr, prefix)

  # Prep Heatmap
  g_trans_mat <- matrix(0, nrow = ape::Nedge(tr), ncol = length(g_trans_edges))

  for (i in 1:length(g_trans_edges)) {
    g_trans_mat[, i] <- g_trans_edges[[i]]
    g_trans_mat[geno_confidence[[i]] == 0, i] <- NA
  }

  p_trans_edges[tr_and_pheno_hi_conf == 0] <- -1
  p_mat <- matrix(p_trans_edges, nrow = length(p_trans_edges), ncol = 1)
  colnames(p_mat) <- "Edge Type"
  phenotype_annotation <- as.data.frame(p_mat)
  row.names(phenotype_annotation) <- 1:nrow(phenotype_annotation)

  temp_g_trans_mat <- cbind(phenotype_annotation, g_trans_mat)
  g_trans_mat <-
    temp_g_trans_mat[order(temp_g_trans_mat[, 1],
                           na.last = FALSE, decreasing = FALSE ),
                     2:ncol(temp_g_trans_mat),
                     drop = FALSE]
  colnames(g_trans_mat) <- row.names(trans_hit_vals)

  significant_loci <-
    data.frame("Locus Significance" = rep("Not Significant", ncol(g_trans_mat)),
               stringsAsFactors = FALSE)
  row.names(significant_loci) <- row.names(trans_hit_vals)
  log_p_value <- data.frame(trans_hit_vals)
  significant_loci[log_p_value > fdr] <- "Significant"

  if (!is.null(snp_in_gene)) {
    snp_in_gene <- as.data.frame(snp_in_gene, row.names = 1)
    colnames(snp_in_gene) <- "Variants in Group"
    snp_in_gene <-
      snp_in_gene[row.names(snp_in_gene) %in% row.names(log_p_value), ,
                  drop = FALSE]
    column_annot <- cbind(significant_loci, log_p_value, snp_in_gene)
  } else {
    column_annot <- cbind(significant_loci, log_p_value)
  }

  ordered_by_p_val <-
    g_trans_mat[, match(row.names(log_p_value)[order(log_p_value[, 1])],
                         colnames(g_trans_mat)),
                 drop = FALSE]
  column_annot_ordered_by_p_val <-
    column_annot[match(row.names(log_p_value)[order(log_p_value[, 1])],
                       row.names(column_annot)), , drop = FALSE]
  if (ncol(column_annot_ordered_by_p_val) == 2) {
    colnames(column_annot_ordered_by_p_val) <- c("Locus Significance",
                                                 "-ln(FDR Corrected P-value)")
  } else {
    colnames(column_annot_ordered_by_p_val) <- c("Locus Significance",
                                                 "-ln(FDR Corrected P-value)",
                                                 "Variants in Group")
  }

  phenotype_annotation[phenotype_annotation == -1] <- "Low Confidence"
  phenotype_annotation[phenotype_annotation == 0] <- "Non-Transition"
  phenotype_annotation[phenotype_annotation == 1] <- "Phenotype Transition"

  ann_colors <- list(`Locus Significance` = c(`Not Significant` = "white",
                                              Significant = "blue"),
                     `Edge Type` = c(`Low Confidence` = "grey",
                                     `Non-Transition` = "white",
                                     `Phenotype Transition` = "red",
                                     `Genotype Transition` = "black"))

  can_be_plotted <- check_if_g_mat_can_be_plotted(ordered_by_p_val)
  if (can_be_plotted) {
    cell_width_value <- image_width / ncol(ordered_by_p_val)
    colnames(ordered_by_p_val) <- substr(colnames(ordered_by_p_val), 1, 20)
    # End Heatmap Prep

    # Plot Heatmap with Genotype and Phenotype Transitions Per Tree Edge
    pheatmap::pheatmap( # Plot the heatmap
      ordered_by_p_val,
      legend = FALSE,
      main = "Synchronous Test: Rows=Tree Edges, Columns=Genotypes",
      cluster_cols = TRUE,
      na_col = "grey",
      cluster_rows  = FALSE,
      show_rownames = FALSE,
      show_colnames = FALSE,
      color = c("white", "black"),
      annotation_col = column_annot_ordered_by_p_val,
      annotation_row = phenotype_annotation,
      annotation_colors = ann_colors,
      fontsize = 8,
      angle_col = 45,
      cellwidth = cell_width_value)
  }

  # Plot Phenotype Transitions on Tree
  pheno_conf_as_list <- list(tr_and_pheno_hi_conf)
  p_trans_edges_as_list <- list(p_trans_edges)
  graphics::par(mfrow = c(1, 1),
                mgp   = c(3, 1, 0),
                oma   = c(0, 0, 4, 0),
                mar = c(4, 4, 4, 4),
                xpd = FALSE)
  plot_tr_w_color_edges(tr,
                        p_trans_edges_as_list,
                        pheno_conf_as_list,
                        "grey",
                        "red",
                        paste0("\n Phenotype"),
                        "recon",
                        1,
                        "No transition",
                        "Transition",
                        tr_type,
                        strain_key = strain_key,
                        tip_name_log = TRUE)

  # Prep data to report p-value rank
  rank_mat <- trans_hit_vals
  rank_mat[rank_mat == 0] <- NA
  genotype_rank_order <- rank(1 / rank_mat[, 1],
                              na.last = TRUE,
                              ties.method = "first")

  # Loop through significant hits:
  for (j in genotype_rank_order) {
    if (trans_hit_vals[j, 1] > fdr) {
      graphics::par(mfrow = c(1, 1),
                    mgp   = c(3, 1, 0),
                    oma   = c(0, 0, 4, 0),
                    mar = c(4, 4, 4, 4),
                    xpd = FALSE)

      # Plot Genotype Reconstruction on Tree
      plot_tr_w_color_edges(tr,
                            g_trans_edges,
                            geno_confidence,
                            "grey",
                            "red",
                            paste0(row.names(trans_hit_vals)[j],
                                   "\n Genotype transitions"),
                            "recon",
                            j,
                            "No transition",
                            "Transition",
                            tr_type,
                            strain_key = NULL,
                            tip_name_log = TRUE)

      # Plot Null Distribution Histogram
      p_val_formated <- formatC(trans_hit_vals[j, 1], format = "e", digits = 1)
      p_val_rank_formated <- rank(1 / rank_mat,
                                  na.last = TRUE,
                                  ties.method = "average")[j]
      title_line_one <- row.names(trans_hit_vals)[j]
      # title_line_one <- expression(paste("N"["Synchronous"],
      #                                    " Null Distribution"))
      title_line_two <- bquote(paste("-ln(FDR Corrected P-value) = ",
                                    .(p_val_formated),
                                    " P-value rank = ",
                                    .(p_val_rank_formated)))

      max_x <- max(trans_perm_obs_results$permuted_count[[j]],
                   trans_perm_obs_results$observed_overlap[j])
      graphics::hist(trans_perm_obs_results$permuted_count[[j]],
           xlim = c(0, max_x),
           col = transparent_teal,
           ylab = "Count",
           xlab = expression("N"["Synchronous"]),
           main = title_line_one)
      graphics::abline(v = trans_perm_obs_results$observed_overlap[j],
                       col = transparent_red,
                       lwd = 4)
      graphics::mtext(title_line_two, side = 3, cex = .6 * hist_cex_size)
      graphics::legend("topright",
             legend = c("Null", "Observed"),
             col = c(transparent_teal, transparent_red),
             pch = 15,
             bty = "n",
             cex = hist_cex_size)
    }
  }
  grDevices::dev.off()
}

#' Plot all results from the Continuous Test
#'
#' @details (1) Plot all genotype's -ln(FDR corrected P-value) on a Manhattan
#'   plot. (2) Plot a heatmap with genotypes in the columns and tree edges in
#'   the rows.|Delta Phenotype| plotted as a row annotation. P-value,
#'   user-defined  significance, and number of variants in group (if grouped)
#'   plotted as column annotations. (3) Plot the continuous phenotype. (4) If
#'   any hits are significant after FDR correction, then plot histograms & trees
#'   for each of the significant loci.
#'
#' @noRd
#' @param disc_cont String, should be "continuous."
#' @param tr Phylo.
#' @param fdr Number. False discovery rate.
#' @param dir Character. Output path.
#' @param name Character. Output name.
#' @param pval_all_transition List of two data.frames.
#'   \describe{
#'     \item{hit_pvals}{Dataframe. 1 column. Nrow = number of genotypes.
#'     Row.names = genotypes. Column name = "fdr_corrected_pvals". Values
#'     between 1 and 0.}
#'     \item{sig_pvals}{Dataframe. 1 column. Nrow = number of genotypes that are
#'     significant after FDR correction. Column name =
#'     "fdr_corrected_pvals[fdr_corrected_pvals < fdr]". Row.names = genotypes.
#'     Nrow = is variable-- could be between 0 and max number of genotypes. It
#'     will only have rows if the corrected p-value is less than the fdr value.}
#'   }
#' @param pheno_vector Vector. Length = Ntip(tr).
#' @param perm Number.
#' @param results_all_trans List of 6:
#'   \describe{
#'     \item{pvals}{Named numeric vector. Length == number of genotypes. Values
#'     between 1 and 0. Names are genotype names.}
#'     \item{null_intersection}{List of numeric vectors. Length of list ==
#'     number of genotypes. Each vector has length == number of permutations.
#'     Values are integers}
#'     \item{observed_pheno_trans_delta}{List of numeric vectors. Length of
#'     list == number of genotypes. Vectors are of variable length because
#'     length is the number of transition edges for that particular genotype.
#'     Vectors are numeric.}
#'     \item{observed_pheno_non_trans_delta}{List of numeric vectors. Length of
#'     list == number of genotypes. Vectors are of variable length because
#'     length is the number of non-transition edges for that particular
#'     genotype. Vectors are numeric.}
#'     \item{num_genotypes}{Integer. The number of genotypes.}
#'     \item{observed_intersection}{Numeric Vector. Length = number of
#'      genotypes. Integers.}
#'   }
#' @param pheno_anc_rec Vector. The values of the ancestral reconstruction of
#'   the phenotype at each internal node. Length = Nnode(tr).
#' @param geno_reconstruction List of lists. Binary. Number of lists = number of
#'   genotypes. Length(each individual list) == Nedge(tree).
#' @param geno_confidence List. Each entry corresponds to one genotype. Length =
#'   number of genotypes.
#' @param geno_transition Object with two lists: $trans_dir and $transition.
#'   Each list has an entry for each genotype. Each sublist has one value for
#'   each tree edge.
#' @param geno Matrix. Columns = genotypes, rows = samples. Binary.
#' @param pheno_recon_ordered_by_edges Matrix. Dim: nrow = Nedge(tr) x ncol = 2.
#'   Parent (older) node is 1st column. Child (younger) node is the 2nd column.
#'   Ancestral reconstruction value of each node.
#' @param tr_and_pheno_hi_conf List. Length(list) = ncol(mat) == number of
#'   genotypes. Each entry is a vector with length == Nedge(tr). All entries are
#'   0 (low confidence) or 1 (high confidence).
#' @param all_trans_sig_hits Data.frame. 1 column. Colnum names =
#'   "fdr_corrected_pvals". Nrow = variable. Number of genotypes that are
#'   significant after multiple test correction. Values are between 1 and 0.
#'   Rownames are genotypes.
#' @param group_logical Logical. Inidicates whether or not genotypes were
#'   grouped.
#' @param snp_in_gene Either NULL or table of integers where each entry
#'   corresponds to one genotype.
#' @param tr_type Characer. Default = "phylogram". User can supply either:
#'   "phylogram" or "fan." Determines how the trees are plotted in the output.
#'
#' @return Plots continuous test results.
plot_continuous_results <- function(disc_cont,
                                    tr,
                                    fdr,
                                    dir,
                                    name,
                                    pval_all_transition,
                                    pheno_vector,
                                    perm,
                                    results_all_trans,
                                    pheno_anc_rec,
                                    geno_reconstruction,
                                    geno_confidence,
                                    geno_transition,
                                    geno_mat,
                                    pheno_recon_ordered_by_edges,
                                    tr_and_pheno_hi_conf,
                                    all_trans_sig_hits,
                                    group_logical,
                                    snp_in_gene,
                                    tr_type,
                                    strain_key){
  # Check inputs ---------------------------------------------------------------
  check_str_is_discrete_or_continuous(disc_cont)
  check_for_root_and_bootstrap(tr)
  check_tree_is_valid(tr)
  check_is_number(fdr)
  check_if_dir_exists(dir)
  check_is_string(name)
  check_class(pval_all_transition$hit_pvals, "data.frame")
  check_equal(length(pheno_vector), ape::Ntip(tr))
  check_if_permutation_num_valid(perm)
  check_equal(length(results_all_trans$null_intersection), ncol(geno_mat))
  check_equal(length(pheno_anc_rec), ape::Nnode(tr))
  check_equal(length(geno_reconstruction), ncol(geno_mat))
  check_equal(length(geno_reconstruction[[1]]), ape::Nedge(tr))
  check_equal(length(geno_confidence), ncol(geno_mat))
  check_equal(length(geno_transition[[1]]$transition), ape::Nedge(tr))
  check_if_binary_matrix(geno_mat)
  check_dimensions(pheno_recon_ordered_by_edges,
                   ape::Nedge(tr),
                   ape::Nedge(tr),
                   2,
                   2)
  check_equal(length(tr_and_pheno_hi_conf), ncol(geno_mat))
  check_class(group_logical, "logical")
  if (!is.null(snp_in_gene)) {
    if (!is_this_class(snp_in_gene, "table") |
        typeof(snp_in_gene) != "integer") {
      stop("snp_in_gene should be a table of integers")
    }
  }
  check_is_string(tr_type)
  # Function -------------------------------------------------------------------

  # Set up variables
  image_width <- 250
  hist_cex_size <- 1
  trans_edge_mat <- NULL

  # Prep to plot heatmap
  for (i in 1:length(geno_transition)) {
    trans_edge_mat <- cbind(trans_edge_mat, geno_transition[[i]]$transition)
  }
  colnames(trans_edge_mat) <- colnames(geno_mat)

  for (c in 1:ncol(trans_edge_mat)) {
    trans_edge_mat[(1:ape::Nedge(tr))[geno_confidence[[c]] == 0], c] <- NA
  }

  ph_trans <-
    abs(pheno_recon_ordered_by_edges[, 1] - pheno_recon_ordered_by_edges[, 2])

  p_trans_mat <- matrix(ph_trans, nrow = length(ph_trans), ncol = 1)
  colnames(p_trans_mat) <- "|Delta Phenotype|"
  p_trans_mat <- as.data.frame(round(p_trans_mat, 2))

  significant_loci <-
    data.frame("Locus Significance" = rep("Not Significant",
                                          ncol(trans_edge_mat)),
               stringsAsFactors = FALSE)
  row.names(significant_loci) <- colnames(trans_edge_mat)
  significant_loci$Locus.Significance[
    pval_all_transition$hit_pvals[ , 1] > fdr] <- "Significant"
  log_p_value <- data.frame(pval_all_transition$hit_pvals)

  if (!is.null(snp_in_gene)) {
    snp_in_gene <- as.data.frame(snp_in_gene, row.names = 1)

    if (ncol(snp_in_gene) == 2) { # Jelly's bug fix
      row.names(snp_in_gene) <- snp_in_gene[, 1]
      snp_in_gene <- snp_in_gene[, 2, drop = FALSE]
    }

    colnames(snp_in_gene) <- "Variants in Group"
    snp_in_gene <-
      snp_in_gene[row.names(snp_in_gene) %in% row.names(log_p_value), ,
                  drop = FALSE]
    column_annot <- cbind(significant_loci, log_p_value, snp_in_gene)
  } else {
    column_annot <- cbind(significant_loci, log_p_value)
  }

  row.names(p_trans_mat) <- row.names(trans_edge_mat) <- c(1:ape::Nedge(tr))

  ann_colors <- list(
    `Locus Significance` = c(`Not Significant` = "white",
                             `Significant` = "blue"),
    `Genotype Edge` = c(`Low Confidence` = "grey",
                        `Non-Transition` = "white",
                        `Transition` = "black")
  )

  sorted_trans_edge_mat <-
    trans_edge_mat[match(row.names(p_trans_mat)[order(p_trans_mat[, 1])],
                         row.names(trans_edge_mat)), , drop = FALSE]
  ordered_by_p_val <-
    sorted_trans_edge_mat[, match(
      row.names(log_p_value)[order(log_p_value[, 1])],
      colnames(sorted_trans_edge_mat)), drop = FALSE]

  column_annot_ordered_by_p_val <-
    column_annot[match(row.names(log_p_value)[order(log_p_value[, 1])],
                       row.names(column_annot)), , drop = FALSE]

  if (ncol(column_annot_ordered_by_p_val) == 2) {
    colnames(column_annot_ordered_by_p_val) <-
      c("Locus Significance", "-ln(FDR Corrected P-value)")
  } else {
    colnames(column_annot_ordered_by_p_val) <-
      c("Locus Significance", "-ln(FDR Corrected P-value)", "Variants in Group")
  }

  # Create dummy column annotation so that a good heatmap legend prints
  column_annot_ordered_by_p_val <-
    cbind(rep("Non-Transition", nrow(column_annot_ordered_by_p_val)),
          column_annot_ordered_by_p_val)
  colnames(column_annot_ordered_by_p_val)[1] <- "Genotype Edge"
  cell_width_value <- image_width / ncol(ordered_by_p_val)
  colnames(ordered_by_p_val) <- substr(colnames(ordered_by_p_val), 1, 20)
  # End of heatmap prep

  # File name
  if (group_logical) {
    fname <- paste0(dir, "/hogwash_continuous_grouped_", name, ".pdf")
  } else {
    fname <- paste0(dir, "/hogwash_continuous_", name, ".pdf")
  }

  # Start plotting
  grDevices::pdf(fname, width = 8.5, height = 11)
  graphics::par(mfrow = c(1, 1), mar = c(5, 5, 5, 5))

  # Manhattan Plot
  make_manhattan_plot(name,
                      pval_all_transition$hit_pvals,
                      fdr,
                      "continuous")

  # Plot Genotype vs Tree Edge Heatmap

  cluster_col_log <- TRUE
  if (!is.null(snp_in_gene)){
    if (nrow(snp_in_gene) == 1) {
      # Avoid clustering issue if only one group in heatmap
      cluster_col_log <- FALSE
    }
  }

  pheatmap::pheatmap(
    ordered_by_p_val,
    main = "Continuous Test: Rows=Tree Edges, Columns=Genotypes",
    cluster_cols = cluster_col_log,
    na_col = "grey",
    cluster_rows = FALSE,
    legend = FALSE,
    show_rownames = FALSE,
    show_colnames = FALSE,
    color = c("white", "black"),
    annotation_col = column_annot_ordered_by_p_val,
    annotation_row = p_trans_mat,
    annotation_colors = ann_colors,
    angle_col = 45,
    fontsize = 8,
    cellwidth = cell_width_value)

  # Plot Continuous Phenotype on Tree
  graphics::par(mfrow = c(1, 1),
                mgp = c(3, 1, 0),
                oma = c(0, 0, 4, 0),
                mar = c(4, 4, 4, 4),
                xpd = FALSE)
  plot_continuous_phenotype(tr, pheno_vector, pheno_anc_rec, tr_type, strain_key)

  # Only make the following plots for significant loci
  # Prep data to report p-value rank
  rank_mat <- pval_all_transition$hit_pvals
  rank_mat[rank_mat == 0] <- NA
  transparent_grey <- grDevices::rgb(0, 0, 0, 0.25)
  legend_grey <- grDevices::rgb(0, 0, 0, 0.05)
  transparent_red <- grDevices::rgb(1, 0, 0, 0.25)
  transparent_teal <- grDevices::rgb(0.4, 0.6, 1, alpha = 0.5, )
  genotype_rank_order <- rank(1 / rank_mat[, 1],
                              na.last = TRUE,
                              ties.method = "first")

  # Loop through significant hits:
  for (j in genotype_rank_order) {
    if (pval_all_transition$hit_pvals[j, 1] > fdr) {
      graphics::par(mfrow = c(1, 1),
                    mgp   = c(3, 1, 0),
                    oma   = c(0, 0, 4, 0),
                    mar = c(4, 4, 4, 4),
                    xpd = FALSE)
      # Plot Genotype Reconstruction on the Tree

      # Plot Genotype Transitions on the Tree
      plot_tr_w_color_edges(tr,
                            geno_transition,
                            geno_confidence,
                            "grey",
                            "red",
                            paste0(row.names(pval_all_transition$hit_pvals)[j],
                                  "\n Genotype transition"),
                            "trans",
                            j,
                            "No transition",
                            "Transition",
                            tr_type,
                            strain_key = NULL,
                            tip_name_log = TRUE)
      # Histogram of the |delta phenotype| for only high conf. edges, colored by
      #   genotype non-transition vs transition edges.
      graphics::par(mfrow = c(2, 2),
                    mgp   = c(3, 1, 0),
                    oma   = c(0, 0, 4, 0),
                    mar = c(4, 4, 4, 4),
                    xpd = FALSE)
      hist_abs_hi_conf_delta_pheno(results_all_trans,
                                   tr,
                                   j,
                                   transparent_grey,
                                   transparent_red)
      # Plot Null Distribution of intersection
      p_val_formated <- formatC(pval_all_transition$hit_pvals[j, 1],
                                format = "e",
                                digits = 1)
      p_val_rank_formated <- rank(1 / rank_mat, na.last = TRUE, ties.method = "average")[j]
      title_line_one <- row.names(pval_all_transition$hit_pvals)[j]
      # title_line_one <- expression(paste("N"["Continuous"],
      #                                    " Null Distribution"))
      title_line_two <- bquote(paste("-ln(FDR Corrected P-value) = ",
                                     .(p_val_formated),
                                     " P-value rank = ",
                                     .(p_val_rank_formated)))
      x_min <- 0
      x_max <- max(as.numeric(results_all_trans$null_intersection[[j]]),
                   results_all_trans$observed_intersection[[j]])

      graphics::hist(log(results_all_trans$null_intersection[[j]]),
                     col = transparent_teal,
                     main = title_line_one,
                     cex.main = hist_cex_size,
                     cex.lab = hist_cex_size,
                     cex.axis = hist_cex_size,
                     ylab = "Count",
                     xlab = expression("N"["Continuous"]),
                     xlim = c(x_min, x_max))
      graphics::mtext(title_line_two, side = 3, cex = .6 * hist_cex_size)
      graphics::abline(v =
                         as.numeric(results_all_trans$observed_intersection[j]),
                       col = transparent_red,
                       lwd = 4)
      graphics::legend("topright",
             legend = c("Null", "Observed"),
             col = c(transparent_teal, transparent_red),
             pch = 15,
             bty = "n",
             cex = hist_cex_size)

    }
  }

  grDevices::dev.off()
  # End Plotting

  # Prep output data
  trans_dir_edge_mat <- NULL
  for (i in 1:length(geno_transition)) {
    trans_dir_edge_mat <- cbind(trans_dir_edge_mat,
                                geno_transition[[i]]$trans_dir)
  }

  colnames(trans_dir_edge_mat) <- colnames(geno_mat)

  for (c in 1:ncol(trans_dir_edge_mat)) {
    trans_dir_edge_mat[(1:ape::Nedge(tr))[geno_confidence[[c]] == 0], c] <- NA
  }

  all_tables <- all_lists <- rep(list(NULL), ncol(trans_dir_edge_mat))
  delta_pheno_table <- matrix(0, nrow = 3, ncol = 1)
  row.names(delta_pheno_table) <- c("geno_parent_0_child_1",
                                    "geno_parent_1_child_0",
                                    "geno_no_change")
  colnames(delta_pheno_table) <- c("sum(|delta_phenotype|)")
  for (i in 1:ncol(trans_dir_edge_mat)) {
    temp_table <- delta_pheno_table
    temp_table[1, 1] <-
      sum(p_trans_mat[which(trans_dir_edge_mat[, i] == 1),  1], na.rm = TRUE)
    temp_table[2, 1] <-
      sum(p_trans_mat[which(trans_dir_edge_mat[, i] == -1), 1], na.rm = TRUE)
    temp_table[3, 1] <-
      sum(p_trans_mat[which(trans_dir_edge_mat[, i] == 0),  1], na.rm = TRUE)
    all_tables[[i]] <- temp_table
  }
  names(all_tables) <- colnames(geno_mat)
  delta_pheno_table <- all_tables
  delta_pheno_list <- rep(list(0), 3)
  names(delta_pheno_list) <- c("geno_parent_0_child_1",
                               "geno_parent_1_child_0",
                               "geno_no_change")
  for (i in 1:ncol(trans_dir_edge_mat)) {
    temp_list <- delta_pheno_list
    temp_list[[1]] <- p_trans_mat[which(trans_dir_edge_mat[, i] == 1),  1]
    temp_list[[2]] <- p_trans_mat[which(trans_dir_edge_mat[, i] == -1), 1]
    temp_list[[3]] <- p_trans_mat[which(trans_dir_edge_mat[, i] == 0),  1]
    all_lists[[i]] <- temp_list
    names(all_lists)[i] <- colnames(trans_dir_edge_mat)[i]
  }
  delta_pheno_list <- all_lists

  # Return output --------------------------------------------------------------
  results <- list("trans_dir_edge_mat" = trans_dir_edge_mat,
                  "p_trans_mat" = p_trans_mat,
                  "delta_pheno_table" = delta_pheno_table,
                  "delta_pheno_list" = delta_pheno_list )
  return(results)
}

#' Assign colors to each user supplied strain type
#'
#' @param strain_key Matrix. 1 column of user supplied characters describing strain type.
#'
#' @return strain_color_key. Matrix. 2 columns. 1st column = user supplied
#'   characters describing strain type. 2nd column, "color_code" is hex code for plotting color.
#' @noRd
assign_colors_to_strain_types <- function(tree, strain_key){
  unique_strain_names <- unique(strain_key[, 1])
  color_hexcodes <- rainbow(n = length(unique_strain_names))
  strain_key <- match_order_to_tree_tips(tree, strain_key)

  # initialize a column that will hold color hex code
  strain_color_key <- cbind(strain_key, rep(NA, nrow(strain_key)))
  for (i in 1:nrow(strain_color_key)) {
    for (j in 1:length(unique_strain_names)) {
      if (strain_color_key[i, 1] == unique_strain_names[j]) {
        strain_color_key[i, 2] <- color_hexcodes[j]
      }
    }
  }
  colnames(strain_color_key)[2] <- "color_code"
  return(strain_color_key)
}
